test/e2e/storage/csimock/csi_node_volume_stats.go

```diff
@@ -0,0 +1,218 @@
+						expectedError:  codes.OK,
+					},
+				},
+				nodeVolumeStatRequired:      true,
+				nodeVolumeConditionRequired: true,
+				nodeAbnormalVolumeCondition: false,
+			},
+			{
+				name: "return normal volume stats without volume condition",
+				expectedCalls: []csiCall{
+					{
+						expectedMethod: "NodeGetVolumeStats",
+						expectedError:  codes.OK,
+					},
+				},
+				nodeVolumeStatRequired:      true,
+				nodeVolumeConditionRequired: false,
+				nodeAbnormalVolumeCondition: false,
+			},
+			{
+				name: "return normal volume stats with abnormal volume condition",
+				expectedCalls: []csiCall{
+					{
+						expectedMethod: "NodeGetVolumeStats",
+						expectedError:  codes.OK,
+					},
+				},
+				nodeVolumeStatRequired:      true,
+				nodeVolumeConditionRequired: true,
+				nodeAbnormalVolumeCondition: true,
+			},
+		}
+		for _, t := range tests {
+			test := t
+			ginkgo.It(test.name, func(ctx context.Context) {
+				ginkgo.By(fmt.Sprintf("volume stats: %+v", test))
+				// Hooks appear to be required for enableNodeVolumeStat.
+				m.init(ctx, testParameters{
+					registerDriver:            true,
+					enableNodeVolumeStat:      test.nodeVolumeStatRequired,
+					enableNodeVolumeCondition: test.nodeVolumeConditionRequired,
+					hooks:                     createGetVolumeStatsHook(test.nodeAbnormalVolumeCondition),
+				})
+				ginkgo.DeferCleanup(m.cleanup)
+				_, claim, pod := m.createPod(ctx, pvcReference)
+				if pod == nil {
+					return
+				}
+				// Wait for PVC to get bound to make sure the CSI driver is fully started.
+				err := e2epv.WaitForPersistentVolumeClaimPhase(ctx, v1.ClaimBound, f.ClientSet, f.Namespace.Name, claim.Name, time.Second, framework.ClaimProvisionTimeout)
+				framework.ExpectNoError(err, "while waiting for PVC to get provisioned")
+
+				ginkgo.By("Waiting for pod to be running")
+				err = e2epod.WaitForPodNameRunningInNamespace(ctx, m.cs, pod.Name, pod.Namespace)
+				framework.ExpectNoError(err, "Failed to start pod: %v", err)
+				ginkgo.By("Waiting for all remaining expected CSI calls")
+				err = wait.PollUntilContextTimeout(ctx, time.Second, csiNodeVolumeStatWaitPeriod, true, func(c context.Context) (done bool, err error) {
+					var index int
+					_, index, err = compareCSICalls(ctx, trackedCalls, test.expectedCalls, m.driver.GetCalls)
+					if err != nil {
+						return true, err
+					}
+					if index == 0 {
+						// No CSI call received yet
+						return false, nil
+					}
+					if len(test.expectedCalls) == index {
+						// all calls received
+						return true, nil
+					}
+					return false, nil
+				})
+				if test.nodeVolumeStatRequired {
+					// try to use ```csi.NewMetricsCsi(pv.handler).GetMetrics()``` to get metrics from csimock driver but failed.
+					// the mocked csidriver register doesn't regist itself to normal csidriver.
+					if test.nodeVolumeConditionRequired {
+						nodeName, err := e2epod.GetPodNodeName(ctx, f.ClientSet, pod.Namespace, pod.Name)
```

Comment:
Why do we need this new function? can we just not use `pod.Spec.NodeName` or something similar?

Reply:
The ```cli.Pods().Create()``` method returns this pod, and the nodeName field might be missing at that time.